File: frontend/src/types/billing.ts
export interface TimePeriod {
  start: string;
  end: string;
}

export interface Metrics {
  amount: string;
  unit: string;
}

export interface GroupMetrics {
  BlendedCost?: Metrics;
  UnblendedCost?: Metrics;
  UsageQuantity?: Metrics;
}

export interface Group {
  keys: string[];
  metrics: GroupMetrics;
}

export interface ResultByTime {
  time_period: TimePeriod;
  total?: GroupMetrics;
  groups: Group[];
  estimated: boolean;
}

export interface CostDataResponse {
  time_period: TimePeriod;
  granularity: string;
  group_by: Array<{ Type: string; Key: string }>;
  results: ResultByTime[];
  dimension_key?: string;
  next_page_token?: string;
}

export interface HealthResponse {
  status: string;
  timestamp: string;
  aws_config: boolean;
}

export interface FilterState {
  startDate: string;
  endDate: string;
  granularity: 'DAILY' | 'MONTHLY';
  groupByDimension?: string;
  metrics: string[];
  serviceFilter?: string;
  regionFilter?: string;
  chargeType?: string;
  includeSupport: boolean;
  includeOtherSubscription: boolean;
  includeUpfront: boolean;
  includeRefund: boolean;
  includeCredit: boolean;
  includeRiFee: boolean;
}

File: backend/app/models/billing.py
from pydantic import BaseModel
from typing import List, Dict, Optional, Any
from datetime import datetime
from .credentials import AWSCredentials


class TimePeriod(BaseModel):
    start: str
    end: str


class Metrics(BaseModel):
    amount: str
    unit: str


class GroupMetrics(BaseModel):
    BlendedCost: Optional[Metrics] = None
    UnblendedCost: Optional[Metrics] = None
    UsageQuantity: Optional[Metrics] = None


class Group(BaseModel):
    keys: List[str]
    metrics: GroupMetrics


class ResultByTime(BaseModel):
    time_period: TimePeriod
    total: Optional[GroupMetrics] = None
    groups: List[Group] = []
    estimated: bool = False


class CostDataRequest(BaseModel):
    credentials: AWSCredentials
    time_period: TimePeriod
    granularity: str = "DAILY"
    group_by: List[Dict[str, str]] = []
    metrics: List[str] = ["BlendedCost"]
    filter: Optional[Dict[str, Any]] = None


class CostDataResponse(BaseModel):
    time_period: TimePeriod
    granularity: str
    group_by: List[Dict[str, str]]
    results: List[ResultByTime]
    dimension_key: Optional[str] = None
    next_page_token: Optional[str] = None


class DimensionRequest(BaseModel):
    credentials: AWSCredentials
    dimension: str
    time_period: TimePeriod


class AccountInfoRequest(BaseModel):
    credentials: AWSCredentials


class HealthResponse(BaseModel):
    status: str
    timestamp: datetime